Enable fraud proof processing in the domain block processor and add tests

NingLin-P opened the pull request:
This PR enables fraud proof processing in the domain block processor and adds tests for invalid state transition proof creation and verification, it covers both `initialize_block`, `apply_extrinsic`, and `finalize_block` proof, but they are ignored due to fraud proof creation being disabled for executor currently.

This PR also adds digest for `apply_extrinsic`, and `finalize_block` proofs when generating them, otherwise, they will be invalid proofs as the test shows.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

NingLin-P committed:
Add digest when generate invalid state transition proof

Actually apply_extrinsic/finalize_block proof also need digest to generate valid proof

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add invalid state transition proof creation/varification tests

This commit add tests for initialize_block proof, apply_extrinsic proof, and finalize_block proof,
they are ignored due to currently fraud proof creation is disabled for executor.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Enable test fraud_proof_verification_in_tx_pool_should_work

Enable this test since we can handle fraud proof call_data properly now,
this commit also further ensure that the test will not produce a fraud proof
that target the block 1.

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on domains/client/domain-executor/src/tests.rs:
```diff
@@ -80,6 +81,193 @@ async fn test_executor_full_node_catching_up() {
+    let original_submit_bundle_tx = bundle_to_tx(bundle.clone().unwrap());
+    let bad_submit_bundle_tx = {
+        let mut signed_opaque_bundle = bundle.unwrap();
+        for rc in signed_opaque_bundle.bundle.receipts.iter_mut() {
```
`rc` hurts readability, please just use `receipt`, it's not that long.

liuchengxu commented on domains/client/domain-executor/src/tests.rs:
```diff
@@ -80,10 +81,194 @@ async fn test_executor_full_node_catching_up() {
+#[ignore]
+async fn test_finalize_block_proof_creation_and_verification_should_work() {
+    test_invalid_state_transition_proof_creation_and_verification(2).await
+}
```
When will these tests be enabled? What features to enable them are still missing?

NingLin-P replied:
When we enable fraud proof here:
https://github.com/subspace/subspace/blob/1ab1eab738cf7a73f1d3c127d8e0b61c9e1904de/domains/client/domain-executor/src/domain_block_processor.rs#L385-L387

It seems we are good to proceed with this step as https://github.com/subspace/subspace/issues/1230 is completed, right?

liuchengxu replied:
Yes, it's time to enable the fraud proof again, especially with these tests added now.

liuchengxu reviewed (commented):
Make sense overall.

NingLin-P committed:
Enable fraud proof processing in domain block processor

This commit also enables the fraud proof tests and come with some minor refactorings

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Apply review suggestion

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Merge branch 'main' into fraud-proof-test

NingLin-P committed:
Merge branch 'main' into fraud-proof-test
